abci.md fixup (#339)
tendermint/spec: spec/abci/abci.md

The `Echo`, `Info`, `InitChain`, `BeginBlock`, `EndBlock`, `Commit` methods
do not return errors. An error in any of these methods represents a critical
issue that Tendermint has no reasonable way to handle. The problem should be
addressed and both Tendermint and the application should be restarted.

## Events

Some methods (`CheckTx, BeginBlock, DeliverTx, EndBlock`)
include an `Events` field in their `Response*`. Each event contains a type and a
list of attributes, which are key-value pairs denoting something about what happened
during the method's execution.
The `CheckTx`, `BeginBlock`, `DeliverTx`, `EndBlock` methods include an `Events`
field in their `Response*`. Each event contains a `type` and a list of `EventAttribute`s,
which are key-value pairs denoting something about what happened during the method's execution.

Events can be used to index transactions and blocks according to what happened
`Event`s can be used to index transactions and blocks according to what happened
during their execution. Note that the set of events returned for a block from
`BeginBlock` and `EndBlock` are merged. In case both methods return the same

> tag, only the value defined in `EndBlock` is used.

Author: This is a great question that I do not know the answer to. I had assumed that, based on the statement on line 71, that the merging was really an insertion into the set that overwrote the previous value. Before I go spelunking through the event indexing code to find out, perhaps @alexanderbez may know the answer here?

> tag, only the value defined in `EndBlock` is used.

Author: https://github.com/tendermint/spec/issues/345

## Errors

The `Query`, `CheckTx` and `DeliverTx` methods include a `Code` field in their `Response*`.
This field is meant to contain an application-specific response code.
A response code of `0` indicates no error.  Any other response code
indicates to Tendermint that an error occurred.

These methods also return a `Codespace` string to Tendermint. This field is
used to disambiguate `Code` values returned by different domains of the
application. The `Codespace` is a namespace for the `Code`.

The `Echo`, `Info`, `InitChain`, `BeginBlock`, `EndBlock`, `Commit` methods
do not return errors. An error in any of these methods represents a critical
issue that Tendermint has no reasonable way to handle. If there is an error in one
of these methods, the application must crash to ensure that the error is safely
handled by an operator.

## Events

The `CheckTx`, `BeginBlock`, `DeliverTx`, `EndBlock` methods include an `Events`
field in their `Response*`. Each event contains a `type` and a list of `EventAttribute`s,

> which are key-value pairs denoting something about what happened during the method's execution.

Author: There does appear to be this separate page for application development that exists within the [docs folder](https://docs.tendermint.com/master/app-dev/indexing-transactions.html) within tendermint as well as this page on [subscribing to events](https://docs.tendermint.com/master/tendermint-core/subscription.html). None of these seem to exactly define 'what' an event is or how they should be used. I'll add a note that these do not impact consensus in any way here.

> `Event` values can be used to index transactions and blocks according to what happened

Reviewer: ![events](https://user-images.githubusercontent.com/27455974/131698102-0246eb85-a171-4e9c-84e8-2ad2fd1198bf.jpg)

I find adding a schematic to the descriptions will help with the more visual learners

Author: I think it's not a bad call, but I'd prefer to may do a bit more polishing if adding an image. My goal with this PR was to spot-fix a few issues or ambiguities I noticed in the spec as it related to issue #316.

during their execution. Note that the set of events returned for a block from
`BeginBlock` and `EndBlock` are merged. In case both methods return the same
key, only the value defined in `EndBlock` is used.

Each event has a `type` which is meant to categorize the event for a particular
`Response*` or `Tx`. A `Response*` or `Tx` may contain multiple events with duplicate
`type` values, where each distinct entry is meant to categorize attributes for a

> particular event. Every key and value in an event's attributes must be UTF-8

Author: @marbar3778  Do you know the answer to this?

> particular event. Every key and value in an event's attributes must be UTF-8

Author: https://github.com/tendermint/spec/issues/346

## EvidenceType

Tendermint's security model relies on the use of "evidence". Evidence is proof of
malicious behaviour by a network participant. It is the responsibility of Tendermint
to detect such malicious behaviour. When malicious behavior is detected, Tendermint
will gossip evidence of the behavior to other nodes and commit the evidence to 

> the chain and once it is verified by all validators. This evidence will then be 

Reviewer: ```suggestion
the chain once it is verified by all validators. This evidence will then be 
```

State sync allows new nodes to rapidly bootstrap by discovering, fetching, and applying
state machine snapshots instead of replaying historical blocks. For more details, see the
[state sync section](../spec/p2p/messages/state-sync.md).

New nodes will discover and request snapshots from other nodes in the P2P network.
A Tendermint node that receives a request for snapshots from a peer will call

> `ListSnapshots` on its application to retrieve any local state snapshots. In addition,
> the new node will offer each snapshot received from a peer to its local application

Reviewer: ```suggestion
`ListSnapshots` on its application to retrieve any local state snapshots. After receiving snapshots from peers,
the new node will offer each snapshot to its local application
```

    | Name   | Type                      | Description                         | Field Number |
    |--------|---------------------------|-------------------------------------|--------------|
    | events | repeated [Event](#events) | type & Key-Value events for indexing | 1           |

* **Usage**:
    * Signals the beginning of a new block.
    * Called prior to any `DeliverTx` method calls.
    * The header contains the height, timestamp, and more - it exactly matches the
    Tendermint block header. We may seek to generalize this in the future.
    * The `LastCommitInfo` and `ByzantineValidators` can be used to determine
    rewards and punishments for the validators. **NOTE:** validators here do not

>     include public keys.

Author: Removed.

* **Usage**:

    * Technically optional - not involved in processing blocks.
    * Guardian of the mempool: every node runs `CheckTx` before letting a
    transaction into its local mempool.
    * The transaction may come from an external user or another node
    * `CheckTx` need not execute the transaction in full, but rather a light-weight
    yet stateful validation, like checking signatures and account balances, but

>     not running code in a virtual machine.

Reviewer 2: It need not, but I think we perhaps want a stronger statement: We seem to _rely_ on the fact that checking a transaction doesn't actually apply any state changes. If that's correct, then "stateful" here may be doing a bit more work than it should. What I think this is trying to say is something like

> `CheckTx` validates the transaction against the current state of the application, for example checking signatures and account balances, but does not apply any state changes described in the transaction.

Does that seem right? Presumably the application could actually run the transaction against a non-persistent clone of the current state to see if it's valid (an implementation detail) but it seems like "non-persistent" is important.

## Data Types

> The data types not listed below are the same as the [core data structures](../spec/core/data_structures.md). The ones listed below have specific changes to better accommodate applications.

Reviewer 2: This seems confusing, I'm not sure how the reader is expected to know which types are supposed to be here vs. not here ("anyone who is absent, please raise your hand"). 

What I _think_ this is saying is something like:

> Most of the data structures used in ABCI are shared [common data structures](../spec/core/data_structures.md) with Tendermint Core. In certain cases, however, ABCI uses different structures, which are documented here:
